bin/version-check.js

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env node
+/** @format */
+/* eslint-disable no-console, no-process-exit */
+
+/**
+ * Checks the user's node and npm versions to make sure the major
+ * versions match. This script will exit with a non-zero status
+ * (causing the build to fail) if the major version of `node` and
+ * `npm` don't match.
+ */
+
+const chalk = require( 'chalk' );
+const checkNodeVersion = require( 'check-node-version' );
+const semver = require( 'semver' );
+const packageJSON = require( '../package.json' );
+
+const majorNodeVersion = semver.major( packageJSON.engines.node );
+const majorNpmVersion = semver.major( packageJSON.engines.npm );
+
+function buildMessage( { engine, engines, results } ) {
```

Comment:
would we ever call this function without one or more of these arguments?
would we call it from outside this module?

there's no need to abandon normal JavaScript calling convention.
why not just `buildMessage( engine, engines, result )`?

Reply:
options objects as method of simulating named parameters is a pattern that is becoming more and more popular since es6.  I like it 🤷‍♂️

Reply:
Yeah, it's just what I've been used to, especially on the addons-frontend project. It makes the arguments clearer and doesn't force order as well. True, for this function it's easy to just make them regular arguments, but I thought this API was nicer to read.

Reply:
it changes the runtime semantics and breaks tools designed to interact with JavaScript functions.

```js
arguments.length === 1 // wut? there are three args
```

 - any decent IDE should already autocomplete and hint the params
 - JSDoc is murky for option objects as args when it comes to indicate typing
 - what's wrong with consistency of having an order for arguments? that's a good thing and makes it far easier to search the code than when looking for these objects.
 - it creates needless objects that don't help the program

Reply:
The `arguments.length` thing is definitely weird, though if you know the function takes an object whose params will be de-structured you'd know to run `Object.keys(arguments[0]).length`. But I get why that's *waaaay* less semantic. 😄

Forcing an *order* for arguments is okay until you start getting optional arguments or functions that take many arguments, but I think the issue is that right in the code you can see what each argument is for because it has an accompanying key. This means code review (where an IDE might well not be used) is easier and readability is improved.

I don't think it's about helping the program; it's about helping the developer. It's a nice API to use for a lot of component testing; when testing React components with enzyme you can pass props to components like this: https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/components/TestFlagReview.js#L31

Reply:
> until you start getting optional arguments or functions that take many arguments which can then be hard to reason about

and I would say that's a good time to start adding arguments as objects. 

> right in the code you can see what each argument is for because it has an accompanying key

also handled by IDEs and JSDoc. the problem is JavaScript though - it's inherent in the language that everything is statically indeterminable. you don't have to supply the required number of parameters ever.

> I don't think it's about helping the program; it's about helping the developer.

My argument is that it doesn't help much and at the same time it has major negative consequences that irk me out.

@flootr this is less about you - @samouri and I have been debating this for some time. this is literally one of my biggest code pet-peeves though because there are fewer things we can do subvert the way a language works than ignore its function calling convention

tools are built to support a language, not a project's conventions. if we say we are having trouble knowing what arguments a function wants we should address the problem and consider using a decent IDE.

Reply:
I understand the arguments for sticking with standard arguments (hah, phrasing!) but I would say the experience of using options objects over the past few years hasn't been in issue in terms of maintainability, tool interference, etc. for us on the addons-frontend or addons-linter project.

This particular function doesn't especially benefit from them but the convention I think is helpful. Asking users to get a particular IDE to read our code is a bit hostile IMO and it's also pretty standard, as @samouri mentioned, across the JS community to use this style of arguments. And when not in an editor/IDE (doing code reviews), it can really help to make sense of the intent behind a function call.

I think there's a purity lost here by using option objects, for sure. I just tend to think it's worth it, and I think eventually functions benefit from it when they grow in size or have many default-and-rarely-used arguments. It's also great for dependency injection–preventing the need to remember which private argument needs to be overridden to do DI in a test. The reason I went with it here, even when it isn't needed, is I think consistency across functions is good, so I'd rather do it always than sometimes.

That said, I value consistency a lot so if this is totally non-standard *and* not agreed upon as the way forward for Calypso in the codebase I'll remove it in the string mutation fix patch as well.

Reply:
If we want to make an argument for it then I would prefer we make the _argument_ for it universally. If we have a problem with abandoning how JavaScript passes arguments then we should change it in every function call. Beyond that we're just biting out thumbs at a standard and a convention because we don't like it.

Asking people to use an IDE when they are confused about a function call isn't much different than asking people to avoid one of the most commonly used features of a programming language.

This pattern of passing named arguments is common but only for optional arguments. There's a reason that Python separates `args` and `kwargs` and it's because there's nothing wrong with passing statically-defined arguments in a statically-defined order.

My goodness! This was not a good day for me to enter this discussion so I'm going to back out for a bit. You aren't the reason I'm short and curt in my response. I apologize.

My main point is that we should make the burden-of-proof for justifying a fundamental standard in a programming language higher than basically anything else we do and we should only make that EXTREMELY confusing change when it's absolutely necessary. It's _better_ to abandon JS and compile to it than it is to do stuff like this.

Or, we should also abandon normal scoping rules and code in globally-unique function names because it's incredibly hard outside of an IDE to know which of many identically-named functions are being used… /s